github-actions/create-pr-for-changes/lib/main.ts

```diff
@@ -0,0 +1,204 @@
+import * as core from '@actions/core';
+import {context} from '@actions/github';
+import {createHash} from 'crypto';
+import {existsSync, readFileSync} from 'fs';
+import {GithubConfig, setConfig} from '../../../ng-dev/utils/config.js';
+import {AuthenticatedGitClient} from '../../../ng-dev/utils/git/authenticated-git-client.js';
+import {ANGULAR_ROBOT, getAuthTokenFor, revokeActiveInstallationToken} from '../../utils.js';
+
+const enum ActionResult {
+  created = 'created',
+  failed = 'failed',
+  nothing = 'nothing',
+}
+
+main();
+
+// Helpers
+async function main(): Promise<void> {
+  let authToken: string | null = null;
+
+  try {
+    // Initialize outputs.
+    core.setOutput('result', ActionResult.nothing);
+    core.setOutput('pr-number', '');
+
+    // Set the cached configuration object to be used throughout the action.
+    const config: {github: GithubConfig} = {
+      github: {
+        mainBranchName: 'main',
+        name: context.repo.repo,
+        owner: context.repo.owner,
+      },
+    };
+    setConfig(config);
+
+    // Configure the `AuthenticatedGitClient` to be authenticated with the token for the Angular
+    // Robot.
+    authToken = await getAuthTokenFor(ANGULAR_ROBOT);
+    AuthenticatedGitClient.configure(authToken);
+    /** The authenticated GitClient. */
+    const git = await AuthenticatedGitClient.get();
+    git.run(['config', 'user.email', 'angular-robot@google.com']);
+    git.run(['config', 'user.name', 'Angular Robot']);
+
+    core.info('Initialized git/GitHub client.');
+
+    // Determine which files have been modified.
+    const touchedFiles = git
+      .run(['status', '--porcelain'])
+      .stdout.trim()
+      .split('\n')
+      .filter((l) => l !== '')
+      .map((l) => l.replace(/^\S+\s+/, ''));
+
+    // Check whether there are any changes to commit.
+    if (touchedFiles.length === 0) {
+      // There are no changes to commit/push. Exit.
+      core.info('Skipping PR creation, because there are no changes.');
+      return;
+    } else {
+      core.info(`Found ${touchedFiles.length} affected file(s).`);
+    }
+
+    // Create a branch name based on the provided branch-prefix and the contents of the changed
+    // files.
+    //
+    // NOTE:
+    //   Hashing the contents of the changed files is a heuristic to uniquely-ish represent a
+    //   particular set of changes. It is not 100% accurate (and could result in both false
+    //   positives and false negatives), but should be good enough for our purposes.
+    const branchPrefix = core.getInput('branch-prefix', {required: true});
+    const branchName = `${branchPrefix}-${hashFiles(touchedFiles)}`;
```

Comment:
Why not derive the branch name using the commit sha?

`${branchPrefix}-${git.run(['rev-parse', 'HEAD'])}`

We could then rely on Git to create something that is even more likely to be unique.

Reply:
This is not just about being unique, but about uniquely identifying a set of changes (regardless of other changes that might have happened to the repo).

Imagine that we use this action to modify file `foo.txt`. I can have two separate SHAs that both apply the same change to file `foo.txt` but are different because their parent SHA is different (due to the unrelated file `bar.txt` being modified between the two). In such a case, we want to recognize the both commits apply the same change, even though their SHAs are different.

Does that make sense?

Reply:
Yeah, that makes sense. Will this hold for an action that doesn't always change all of the same files though?  For instance if you have an action that updates all of the generated files, but some of them are unchanged.

Reply:
The current hashing looks at the files modified by the action run and their content. If there are two different runs that result in the same files being touched and the contents of those touched files are the same on both runs, then the two runs will be considered the same.

If you have two runs that result in a different set of files being touched (or if the contents of the touched files has changed between the two runs), then they will be considered different.

This can result in "misses" (i.e. two runs that should be identified as the same being considered different), but that is not a big problem: It just means that we might get two PRs that do the same thing.